lib/auth_processes/mongodb.js

```diff
@@ -1,383 +1,142 @@
+        callback();
       });
-    });
-  };
-};
-
-/*******************************************************************
- *
- * UNIX MIT Kerberos processor
- *
- *******************************************************************/
-const UnixMongoProcessor = function(host, port, service_name, options) {
-  options = options || {};
-  this.host = host;
-  this.port = port;
-
-  // SSIP classes
-  this.Kerberos = require('../kerberos').Kerberos;
-  this.kerberos = new this.Kerberos();
+    }
 
-  // Set up service name
-  service_name = service_name || 'mongodb';
-
-  // Options
-  this.gssapiServiceName = options.gssapiServiceName || service_name;
-  this.gssapiServiceRealm = options.gssapiServiceRealm;
-  this.gssapiCanonicalizeHostName =
-    typeof options.gssapiCanonicalizeHostName === 'boolean'
-      ? options.gssapiCanonicalizeHostName
-      : false;
-
-  // Set up first transition
-  this._transition = UnixMongoProcessor.first_transition(this);
-
-  // Set up target
-  this.target = `${service_name}@${host}`;
-
-  // Number of retries
-  this.retries = 10;
-};
-
-UnixMongoProcessor.prototype.init = function(username, password, callback) {
-  const self = this;
-  this.username = username;
-  this.password = password;
-
-  // Canonicialize host name if needed
-  const performGssapiCanonicalizeHostName = function(gssapiCanonicalizeHostName, host, callback) {
-    if (!gssapiCanonicalizeHostName) return callback();
-
-    // Attempt to resolve the host name
-    dns.resolveCname(host, function(err, r) {
+    // Canonicialize host name if needed
+    performGssapiCanonicalizeHostName(this.canonicalizeHostName, this.host, err => {
       if (err) return callback(err);
 
-      // Get the first resolve host id
-      if (Array.isArray(r) && r.length > 0) {
-        self.host = r[0];
-      }
+      kerberos.initializeClient(self.service, { user: username, password }, (err, client) => {
+        if (err) return callback(err, null);
 
-      callback();
+        self.client = client;
+        callback(null, client);
+      });
     });
-  };
-
-  // Canonicialize host name if needed
-  performGssapiCanonicalizeHostName(this.gssapiCanonicalizeHostName, this.host, function(err) {
-    if (err) return callback(err);
-
-    // Set up target
-    self.target = `${self.gssapiServiceName}@${self.host}`;
+  }
 
-    // Call client initiate
-    self.kerberos.authGSSClientInit(self.target, self.Kerberos.GSS_C_MUTUAL_FLAG, function(
-      err,
-      context
-    ) {
-      self.context = context;
-      // Return the context
-      callback(null, context);
-    });
-  });
-};
+  transition(payload, callback) {
+    if (this._transition == null) {
+      return callback(new Error('Transition finished'));
+    }
 
-UnixMongoProcessor.prototype.transition = function(payload, callback) {
-  if (this._transition == null) return callback(new Error('Transition finished'));
-  this._transition(payload, callback);
-};
+    this._transition(payload, callback);
```

Comment:
This whole transition mechanic looks like another impl of a state machine. Can we add a todo/note to revisit this when we eventually come around to trying to make a single state machine impl?

Reply:
added to a comment in [NODE-1517](https://jira.mongodb.org/browse/NODE-1517)